### ISSUE TITLE
Sik-Ho Tang | Review: Inception-v4 -- Evolved From GoogLeNet, Merged with ResNet Idea (Image Classification).

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#112]
Sik-Ho Tang. [Review: Inception-v4 — Evolved From GoogLeNet, Merged with ResNet Idea (Image Classification)](https://towardsdatascience.com/review-inception-v4-evolved-from-googlenet-merged-with-resnet-idea-image-classification-5e8c339d18bc).

**Comment by NorbertZheng** [comment]
## Overview
In this story, Inception-v4 [1] by Google is reviewed. Inception-v4, evolved from GoogLeNet / Inception-v1, has a more uniform simplified architecture and more inception modules than Inception-v3.

From the below figure, we can see the top-1 accuracy from v1 to v4. And Inception-v4 is better than ResNet.
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/68576b45-b302-4dbd-aade-f08ed232bb32)
Top-1 Accuracy against Number of Operations (Size is the number of parameters).

 - **Inception network with residual connections, an idea proposed by Microsoft ResNet, outperforms similarly expensive Inception network without residual connections.**

With ensemble of 1 Inception-v4 and 3 residual networks, 3.08% error can be achieved in ILSVRC classification task. And this is a 2017 AAAI paper with more than 1000 citations when I was writing this story.

**Comment by NorbertZheng** [comment]
## Brief Overview from Inception-v1 to Inception-v3
Before Starting to talk about Inception-v4, let’s review a bit from v1 to v3 first.
### Inception-v1 / GoogLeNet [2]: Inception Module
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/854c96bd-6351-4021-b859-1ee7dee19f3d)
Inception Module (Left), Inception Module with Dimensionality Reduction (Right).

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/7228ed93-a2c0-48ab-94e9-f34747c8b36b)
Overall Architecture.

Inception module was firstly introduced in Inception-v1 / GoogLeNet. The input goes through 1×1, 3×3 and 5×5 conv, as well as max pooling simultaneously and concatenated together as output. Thus, we don’t need to think of which filter size should be used at each layer.

([My detailed review on Inception-v1 / GoogLeNet](https://medium.com/coinmonks/paper-review-of-googlenet-inception-v1-winner-of-ilsvlc-2014-image-classification-c2b3565a64e7))

**Comment by NorbertZheng** [comment]
Model Ensemble!!! We have multiple paths for pass the information.

**Comment by NorbertZheng** [comment]
### Inception-v2 / BN-Inception [3]: Batch Normalization
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/f72a78b7-0d96-472f-99a2-6a2b37b80877)
Batch Normalization (BN).

 - Batch normalization (BN) was introduced in Inception-v2 / BN-Inception.
 - ReLU is used as activation function to address the saturation problem and the resulting vanishing gradients.

 - But it also makes the output more irregular. **It is advantageous for the distribution of $X$ to remain fixed over time because a small change will be amplified when network goes deeper.**

 - Higher learning rate can be used.

Also, 5×5 conv was replaced by two 3×3 convs for parameter reduction while maintaining the receptive field size.

([My detailed review on Inception-v2 / BN-Inception](https://medium.com/@sh.tsang/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651))

**Comment by NorbertZheng** [comment]
### Inception-v3 [4]: Factorization
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/30959ab6-a737-4d4a-a86d-a3b88a6f301e)
3×3 conv becomes 1×3 and 3×1 convs (Left), 7×7 conv becomes 1×7 and 7×1 convs (Right).

**Factorization was introduced in convolution layer as shown above to further reduce the dimensionality**, so as to reduce the overfitting problem. For example:

By using 3×3 filter, number of parameters = 3×3=9
By using 3×1 and 1×3 filters, number of parameters = 3×1+1×3=6
Number of parameters is reduced by 33%

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/baed04d6-413d-4435-b516-4043fadf915e)
Conventional downsizing (Top Left), Efficient Grid Size Reduction (Bottom Left), Detailed Architecture of Efficient Grid Size Reduction (Right).

And an efficient grid size reduction module was also introduced which is less expensive and still efficient network. With the efficient grid size reduction, say for example in the figure, 320 feature maps are done by conv with stride 2. 320 feature maps are obtained by max pooling. And these 2 sets of feature maps are concatenated as 640 feature maps and go to the next level of inception module.

([My detailed review on Inception-v3](https://medium.com/@sh.tsang/review-inception-v3-1st-runner-up-image-classification-in-ilsvrc-2015-17915421f77c))

**Comment by NorbertZheng** [comment]
## Inception-v4
A more uniform simplified architecture and more inception modules than Inception-v3, is introduced as below:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/272940b2-c740-4b62-880f-c5d116631796)
Inception-v4: Whole Network Schema (Leftmost), Stem (2nd Left), Inception-A (Middle), Inception-B (2nd Right), Inception-C (Rightmost).

This is a pure Inception variant without any residual connections. It can be trained without partitioning the replicas, with memory optimization to backpropagation.

We can see that the techniques from Inception-v1 to Inception-v3 are used. (Batch Normalization is also used but not shown in the figure.)

**Comment by NorbertZheng** [comment]
## Inception-ResNet-v1
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/8cebe01e-a726-496d-9505-07b2bade2852)
Inception-ResNet-v1: Whole Network Schema (Leftmost), Stem (2nd Left), Inception-A (Middle), Inception-B (2nd Right), Inception-C (Rightmost).

By using the above versions of Inception-A, Inception-B and Inception-C, we can have Inception-ResNet-v1. We can see that there is a shortcut connection at the left of each module. This shortcut connection has been a kind of proved that it can help go deeper in ResNet [5].

 - **There is no pooling path, only residual path!!!**

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/e0ca723c-d064-450f-8fdd-b2ce19f3741a)
Inception-Resnet-v1 and Inception-v3.

It has roughly the computational cost of Inception-v3. Inception-Resnet-v1 was training much faster, but **reached slightly worse final accuracy than Inception-v3**.

 - However, **the ReLU used after adding together** makes Inception network not able to go further deeper. **Gradient explosion!!!**
 - In the paper, authors also mentioned that if the number of filters exceeded 1000, the residual variants started to exhibit instabilities, and the network just “died” early during training.

To stabilize the training, **scaling down the residuals before adding them to the previous layer activation**. In general some scaling factors are picked between 0.1 and 0.3 to scale the residuals before being added to the accumulated layer activations.

(Perhaps, if the ReLU is used as pre-activation unit, as mentioned in the improved ResNet using Identity Mapping [6], it may can go much deeper.) (If interest, please visit [my review on Pre-Activation ResNet.](https://towardsdatascience.com/resnet-with-identity-mapping-over-1000-layers-reached-image-classification-bb50a42af03e))

**Comment by NorbertZheng** [comment]
## Inception-ResNet-v2
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/d8a77571-6b0a-470a-bba0-9995e6a347c7)
Inception-ResNet-v2: Inception-A (Leftmost), Inception-B (Middle), Inception-C (Rightmost).

With the whole network schema using the one in Inception-ResNet-v1, Stem using the one in Inception-v4, and the above versions of Inception-A, Inception-B and Inception-C, we can have Inception-ResNet-v2. Again, there is a shortcut connection at the left of each module.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/247ad5af-d436-4789-a262-1bf660a8fbe7)
Inception-Resnet-v2 and Inception-v4.

It has roughly the computational cost of Inception-v4. Inception-ResNet-v2 was training much faster and **reached slightly better final accuracy than Inception-v4**.

 - However, again similarly, if the ReLU is used as pre-activation unit, it may can go much deeper. (If interest, please visit [my review on Improved ResNet.](https://towardsdatascience.com/resnet-with-identity-mapping-over-1000-layers-reached-image-classification-bb50a42af03e))

**Comment by NorbertZheng** [comment]
## Comparison with State-of-the-art Approaches
 - **Multi-crop** (Data Augmentation!!!): is to crop the input image into multi sub-images and input into network for classification, to increase the accuracy. It has been used since AlexNet, VGGNet and so on. To be specific, 10 crop was used in AlexNet, 144 crop was used in GoogLeNet.
 - **Multi-model** (Model Augmentation!!!): is to have multi trained models ensemble together to get a more accurate prediction, just like boosting. It has been used since LeNet, AlexNet and so on.

(Please visit my [AlexNet](https://medium.com/coinmonks/paper-review-of-alexnet-caffenet-winner-in-ilsvrc-2012-image-classification-b93598314160) and [GoogLeNet](https://medium.com/coinmonks/paper-review-of-googlenet-inception-v1-winner-of-ilsvlc-2014-image-classification-c2b3565a64e7) reviews if interested.)

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/6354a00b-b2cd-4b19-82ac-0e70c95e9de8)
Single-Crop Single-Model Results.

 - With Single-Crop Single-Model, Inception-v4 and Inception-ResNet-v2 have the best performance with similar results.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/a682a3e7-1be3-4631-b8b4-10af084a958a)
10/12-Crop Single-Model Results.

 - With 10/12-Crop Single-Model, again, Inception-v4 and Inception-ResNet-v2 again have the best performance with similar results.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/4f888e24-dfa6-4d5d-816a-3be4eb88c450)
144-Crop Single-Model Results.

 - With 144-Crop Single-Model, again, Inception-v4 and Inception-ResNet-v2 again have the best performance with similar results.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/ae2aad4d-0a63-4d0d-be0a-770365110b9a)
144-Crop N-Model Results.

With 144-Crop N-Model, Inception-v4(+Residual), i.e. 1 pure Inception-v4 and 3 Inception-ResNet-v2 models (N=4), has the best performance.

3.1% Top-5 error rate is obtained. This has been a large relative improvement compared with ResNet (3.57% error rate, the winner in ILSVRC 2015) and Inception-v3 (3.58% error rate, 1st Runner Up in ILSVRC 2015).

**Comment by NorbertZheng** [comment]
## References
 - [2017 AAAI] [Inception-v4] [Inception-v4, Inception-ResNet and the Impact of Residual Connections on Learning](https://www.aaai.org/ocs/index.php/AAAI/AAAI17/paper/download/14806/14311).
 - [2015 CVPR] [GoogLeNet / Inception-v1] [Going Deeper with Convolutions](https://www.cs.unc.edu/~wliu/papers/GoogLeNet.pdf).
 - [2015 ICML] [BN-Inception / Inception-v2] [Batch Normalization: Accelerating Deep Network Training by Reducing Internal Covariate Shift](http://proceedings.mlr.press/v37/ioffe15.pdf).
 - [2016 CVPR] [Inception-v3] [Rethinking the Inception Architecture for Computer Vision](https://www.cv-foundation.org/openaccess/content_cvpr_2016/papers/Szegedy_Rethinking_the_Inception_CVPR_2016_paper.pdf).
 - [2016 CVPR] [ResNet] [Deep Residual Learning for Image Recognition](https://www.cv-foundation.org/openaccess/content_cvpr_2016/papers/He_Deep_Residual_Learning_CVPR_2016_paper.pdf).
 - [2016 ECCV] [ResNet with Identity Mapping] [Identity Mappings in Deep Residual Networks](https://arxiv.org/abs/1603.05027).
